app/models/invoice.rb

```diff
@@ -45,7 +47,9 @@ class InvalidInvoiceStateException < RuntimeError; end
   attr_accessible :provider_account, :buyer_account, :friendly_id, :period
 
   validates :provider_account, :buyer_account, :friendly_id, presence: true
+
   validates :period, presence: { :message => 'Billing period format should be YYYY-MM' }
+  validate :year_valid?
```

Comment:
Just to confirm, this will prevent creating/editing any invoice that doesn't meet requirements, but it doesn't interfere with already existing invoices, right?

What if we generate the PDF for an invoice which has an invalid year? that will update the `updated_at` column for the invoice and force the validation, wouldn't that break the PDF generation?

Reply:
Also, I agree this need validation, we can't allow the user to input something like `abcd`

Reply:
With this validation (years within the range) I am pretty sure the existing invoices will meet the requirements (at least in SaaS).
Yeah, and that's why I am concerned about making the rules too restrictive, as we don't want to invalidate existing "pretty much valid" invoices.

Reply:
Well, probably most of invoices are old and are not going to be updated anymore. This would potentially affect a small group of recent invoices which still need some update, and from those, probably there are very few which doesn't meet the requirements, if any.